salesagility/SuiteCRM
# Calculate list price according to cost and utility values

## Description
Added a feature on Quotes modues to use or modify the cost of a product or service and specify an ammount or percentage of utility and calculate the value of List Price according to the cost and utility specified.

1. The procuct/service cost field have been made visible, this one was present but hidden.
2. Add, fields for utility and utility type (amount or percentage) similar to discount and discount type fields.
3. Utility type is like Discount type, a html select element to choose between Amount or Percentage, and depending on the type selected the list price will be calculated in two different ways.
4. Utility is a number, that is expected to represent an amount or a percentage that should be added to the cost in order to obtain the list price.
5. If Utility Type is amount then the list price will be the cost plus the utility.
6. If Utility Type is amount and the utility value is negative it's absolute value must not be greater than the cost, calculateLine function will change the value automaticly
7. If Utility Type is percentage, the list price will be calculated according to this formula:
> ListPrice = Cost /(1-(Utility/100))
- The value of Utility could be negative or positive between -1000 to 90
- Any value greater than 90 will be changed to 90 when calculateLine is executed
8. The field list price is now readonly and calculateLine will calculate its value.
9. All others fields will preserve its function as before.
10. All fields have been reorganized to fit the new fields.
11. Also some lines have been changed only to fit SuiteCRM's coding standards.

## Motivation and Context

Not all the products and services has a static cost all the times, some change from day to day or according to the provider that the company finds every time. Change constantly the cost of the product is difficult and not useful as it not allows a company to compare utility obtain in every sell.
Calculate the list price in a spreadsheet and copy it to quote module is a double process, calculated by hand (even in a calculator) could lead to errors, so leave this function to SuiteCRM could help to automate the process and reduce mistakes, even later on a company could create a report to calculate how much profit a sell or a group of sales have been earned.
 
## How To Test This

Create a few products, assign a cost and a list price.
Create a new quote
Add some products and services
The value of product's cost should be recovered from DB with all the other product's information as before.
Change the product's cost, assign an utility and choose an utility type, as you leave any of this fields you should see how list price and total price changes.
Change utility type between Percentage and Amount, it should change the amount on list price and all the other fields.
Compare the values to be sure that calculations have been made correctly as explained on point 6 and 7.
Save the quote, open it or edit it again, all values should be stored and recovered as you saved it.

## Types of changes


### Final checklist


- [X] I have read the [**How to Contribute**]
(https://suitecrm.com/wiki/index.php/Contributing_to_SuiteCRM) guidelines.
